fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/matchers/PlannerBindings.java

```diff
@@ -21,62 +21,146 @@
     public <T extends Bindable> T get(@Nonnull ExpressionMatcher<T> key) {
         if (bindings.containsKey(key)) {
-            return (T) bindings.get(key);
+            List<? extends Bindable> bindingsForKey = bindings.get(key);
+            if (bindingsForKey.size() == 1) {
+                return (T)bindingsForKey.get(0);
+            } else {
+                throw new NoSuchElementException("attempted to retrieve individual bindable but multiple keys were bound");
+            }
         }
         throw new NoSuchElementException("attempted to extract bindable from binding using non-existent key");
     }
 
     /**
-     * Bind the given key to the provided {@link Bindable}. Note that <code>bindable</code> must be an instance
-     * of the type parameter of <code>key</code>, although the compiler cannot enforce this because of how it is used
-     * in {@link Bindable#bindWithExisting(ExpressionMatcher, PlannerBindings)}.
-     * @param key a matcher with type parameter that is a super class of the class of <code>bindable</code>
-     * @param bindable a bindable object to bind to the key
+     * Retrieve all bindables bound to {@code key} if there is at least one such bindable. The bindables in the returned
+     * list appear in same order as they appear in the list of children of the {@link PlannerExpression} that produced
+     * this set of bindings. If no bindable is bound to this key, throw a {@link NoSuchElementException}.
+     * @param key a matcher
+     * @param <T> the type of {@link Bindable} that was bound to {@code key}
+     * @return a list of bindable objects bound to the key
+     * @throws NoSuchElementException if no bindable objects are bound to the given key
      */
-    public void put(@Nonnull ExpressionMatcher<? extends Bindable> key, @Nonnull Bindable bindable) {
+    @Nonnull
+    @SuppressWarnings("unchecked")
+    public <T extends Bindable> List<T> getAll(@Nonnull ExpressionMatcher<T> key) {
         if (bindings.containsKey(key)) {
-            throw new RecordCoreException("attempted to add a binding that already exists");
+            return (List<T>) bindings.get(key);
+        }
+        throw new NoSuchElementException("attempted to extract bindable from binding using non-existent key");
+    }
+
+    /**
+     * Combine this set of bindings with the given set of bindings. If both sets of bindings contain a binding with the
+     * same key, the resulting set of bindings will contain the <em>concatenation</em> of the bindables from this set
+     * of bindings and the bindables from the given set of bindings, in that order, providing a predictable ordering
+     * of bindings for planner rules that might need it.
+     * @param other a set of bindings, which may share keys with this set of bindings
+     * @return a new set of bindings that contains the bindings from both both sets of bindings
+     */
+    public PlannerBindings mergedWith(@Nonnull PlannerBindings other) {
```

Comment:
Missing nullability annotation on the return type.

Reply:
Added.